cluster-autoscaler/core/static_autoscaler.go

```diff
@@ -665,6 +667,11 @@ func (a *StaticAutoscaler) RunOnce(currentTime time.Time) caerrors.AutoscalerErr
 				scaleDownStatus.Result == scaledownstatus.ScaleDownNoUnneeded) &&
 				a.AutoscalingContext.AutoscalingOptions.MaxBulkSoftTaintCount != 0 {
 				taintableNodes := a.scaleDownPlanner.UnneededNodes()
+				allNodes, err := filterNodeGroups(a.CloudProvider, allNodes...)
+				if err != nil {
+					klog.Warningf("Failed filtering nodes based on selected node groups: %v", err)
+				}
+				// REVIEW: Shall we do the below calls if there's an error? This is essentially gonna mean that untaintableNodes is empty since allNodes is empty.
```

Comment:
Two main comments:
- I don't think the failure mode for err should be "don't clean the taints at all", I think it should be "clean all the nodes". In other words - if the filter returns an error, then I think we should actually use the previous allNodes value. This applies to the previous case as well.
- After these changes taintableNodes used here may "theoretically" be something than is not a strict subset of filtered nodes? Just for sanity check I would limit the taintableNodes to the ones present in filtered nodes as well.

Reply:
> I don't think the failure mode for err should be "don't clean the taints at all", I think it should be "clean all the nodes". In other words - if the filter returns an error, then I think we should actually use the previous allNodes value. This applies to the previous case as well.

Sounds good, going to make the change :+1: 

> After these changes taintableNodes used here may "theoretically" be something than is not a strict subset of filtered nodes? Just for sanity check I would limit the taintableNodes to the ones present in filtered nodes as well.

Yeah "theoretically" the planner should already take into account the selected node groups, but I agree it's a good idea to make sure this is the case. I'll add that.